### DISCUSSION TITLE
Can the dependency on JetBrains.Annotations be removed?

### DISCUSSION THREAD

**Discussion by FriedLychees** [ardalis/GuardClauses#216] [Q&A] [answered]
Since this is such a low level library, it seems a little weird to me that this has a third party dependency.

It would be nice to be able to reference this package in our core code, and not be pulling in a version of Annotations dictated by this package.

JetBrains themselves seem to recommend not referencing Annotations in open-source code (see [How to add](https://blog.jetbrains.com/dotnet/2018/05/03/what-are-jetbrains-annotations/) section).

From a quick scan of the GuardClauses code, it didn't seem to be integral to the functionality. Is it possible that the dependency could be removed?

**Comment by danny-bos-developer** [comment]
Agree

**Comment by ardalis** [comment, marked as answer]
Yes I think it could be removed. I'll create an issue for now to track it.

**Comment by savornicesei** [comment]
Hi!
Is it possible to release this, please?

Thank you.

**Comment by nseguin42** [comment]
I'm a bit confused here. It looks to me like helpful functionality was removed due to a misunderstanding. Rather than remove the annotations, they should have been copied into the source code and marked as `internal` (the officially recommended way for libraries to reference the annotations).

---

[The referenced article](https://blog.jetbrains.com/dotnet/2018/05/03/what-are-jetbrains-annotations/#HowtoaddJetBrainsAnnotationstoourowncodebase?) does not "recommend not referencing Annotations in open-source code" at all. If anything, it gives instructions on how to do it correctly. On the other hand, [this article](https://blog.jetbrains.com/dotnet/2015/08/12/how-to-use-jetbrains-annotations-to-improve-resharper-inspections/) has the statement I believe @FriedLychees misunderstood: 

> "If your code is to be distributed as a NuGet package that is to be consumed by anyone, it’s not appropriate to include a binary dependency on JetBrains.Annotations.dll."

Referencing a binary dependency on `JetBrains.Annotations.dll` is one of two approaches to propagating the annotations to consuming assemblies. As the article explains, the recommended approach for libraries is to include the annotation source code:

> **To use annotations in an assembly that is to be used by third parties:** Include the attributes as source. **DO** define as internal, **DO NOT** generate as conditional. Both you and third party consumers will benefit from the annotations.


---

The functionality that was removed is not super critical to the package (the built-in nullability annotations are good enough). But it was nice for those who use Rider and/or ReSharper.

- [InvokerParameterName] is especially useful in NETSTANDARD/NETFRAMEWORK where [CallerArgumentExpression] is not available. 
- [RegexPattern] and [NoEnumeration] are nice to have everywhere.
- There are some additional attributes that would be useful for this package, e.g. [ContractAnnotation("input:null => halt")] for `Guard.Against.Null`.

**Comment by ardalis** [comment]
Yeah I think I agree. Most of the JetBrains stuff has been happening through community contributions since I don't use ReSharper any longer and am not a Rider user, either. @nseguin42 would you mind creating a new issue (or issues) with specific annotations you would want to see added (back) in and I'll work on that or perhaps we can get community pull requests for them? Maybe it makes sense to include the copy/pasted attribute in the issue description for attributes that should be included in the project.
